File: android/src/main/java/app/capgo/persistent_uuid/PersistentUuidPlugin.java
package app.capgo.persistent_uuid;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "PersistentUuid")
public class PersistentUuidPlugin extends Plugin {

    private PersistentUuid implementation;

    @Override
    public void load() {
        super.load();
        implementation = new PersistentUuid(getContext());
    }

    @PluginMethod
    public void getId(PluginCall call) {
        try {
            PersistentUuid.Result result = implementation.getId(call.getString("scope"));
            call.resolve(toJsObject(result));
        } catch (Exception exception) {
            call.reject("Failed to get persistent UUID", exception);
        }
    }

    @PluginMethod
    public void resetId(PluginCall call) {
        try {
            PersistentUuid.Result result = implementation.resetId(call.getString("scope"));
            call.resolve(toJsObject(result));
        } catch (Exception exception) {
            call.reject("Failed to reset persistent UUID", exception);
        }
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("version", implementation.getPluginVersion());
        call.resolve(ret);
    }

    private JSObject toJsObject(PersistentUuid.Result result) {
        JSObject ret = new JSObject();
        ret.put("id", result.id);
        ret.put("scope", result.scope);
        ret.put("created", result.created);
        return ret;
    }
}

File: ios/Sources/PersistentUuidPlugin/PersistentUuidPlugin.swift
import Foundation
import Capacitor

@objc(PersistentUuidPlugin)
public class PersistentUuidPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "PersistentUuidPlugin"
    public let jsName = "PersistentUuid"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resetId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = PersistentUuid()

    @objc func getId(_ call: CAPPluginCall) {
        do {
            call.resolve(try implementation.getId(call.getString("scope")))
        } catch {
            call.reject("Failed to get persistent UUID: \(error.localizedDescription)")
        }
    }

    @objc func resetId(_ call: CAPPluginCall) {
        do {
            call.resolve(try implementation.resetId(call.getString("scope")))
        } catch {
            call.reject("Failed to reset persistent UUID: \(error.localizedDescription)")
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }
}
